index.js

```diff
@@ -70,63 +72,69 @@ Client.prototype._ref = function () {
   })
 }
 
-Client.prototype._write = function (obj, cb) {
+Client.prototype._write = function (obj, enc, cb) {
   if (this._destroyed) {
-    this.emit('error', new Error('data sent to destroyed Elastic APM client'))
-    return
+    this.emit('error', new Error('write called on destroyed Elastic APM client'))
+    cb()
+  } else if (obj === flush) {
+    if (this._active) {
+      this._onflushed = cb
+      this._chopper.chop()
+    } else {
+      this._chopper.chop(cb)
+    }
+  } else {
+    this._stream.write(obj, cb)
   }
-  return this._stream.write(obj, cb)
 }
 
 Client.prototype.sendSpan = function (span, cb) {
-  return this._write({span}, cb)
+  return this.write({span}, cb)
 }
 
 Client.prototype.sendTransaction = function (transaction, cb) {
-  return this._write({transaction}, cb)
+  return this.write({transaction}, cb)
 }
 
 Client.prototype.sendError = function (error, cb) {
-  return this._write({error}, cb)
+  return this.write({error}, cb)
 }
 
 Client.prototype.flush = function (cb) {
   if (this._destroyed) {
     this.emit('error', new Error('flush called on destroyed Elastic APM client'))
+    if (cb) process.nextTick(cb)
     return
   }
-  if (this._ended) return // TODO: call bacllback?
 
-  if (cb && this._active) {
-    // if there's an outgoing http request, queue the callback to be called
-    // when we're sure that the data has left the system
-    this._onflushed = cb
-    this._chopper.chop()
-  } else {
-    this._chopper.chop(cb)
-  }
+  return this.write(flush, cb)
 }
 
-Client.prototype.end = function (cb) {
+Client.prototype._final = function (cb) {
   if (this._destroyed) {
     this.emit('error', new Error('end called on destroyed Elastic APM client'))
+    cb()
     return
   }
-  if (this._ended) return // TODO: call callback?
-  this._ended = true
   clients[this._index] = null // remove global reference to ease garbage collection
-  if (cb) this.once('finish', cb)
   this._ref()
   this._stream.end()
+  cb()
 }
 
-Client.prototype.destroy = function () {
+// Overwrite destroy instead of using _destroy because readable-stream@2 can't
+// be trusted
```

Comment:
Can you provide some more context here?

Reply:
After a stream is destroyed, we want a call to either `client.write()` or `client.end()` to both emit an error and call the provided callback. Unfortunately, in readable-stream@2 and Node.js <10, this is not consistent. This has been fixed in Node.js 10 and will also be fixed in readable-stream@3.

I've updated the comment to better explain this.

Reply:
Specifically, this is because calling `stream.destroy()` emits `finish` synchronously in Node.js <10 and in readable-stream@2. So when calling `stream.end(callback)` later, the `callback` is attached to the `finish` event, but since it has already fired, nothing happens.

In Node.js 10 and in the upcoming readable-stream@3, this have been fixed so that the `finish` event isn't emitted until you call `stream.end()`.

You can see an isolated test case with the problem here: https://gist.github.com/watson/6afbb67423c979a5c8f311f137184320